File: backend/app/routes/despesas_locais.py
"""
Rotas /despesas-locais — CRUD de despesas fixas/manuais (não importadas via Excel).
Suporta parcelamento (ex: 7/12).
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional

from app.services.storage_service import despesas_locais_storage

router = APIRouter()


class DespesaLocalCreate(BaseModel):
    nome: str
    categoria: Optional[str] = "outros"
    subcategoria: Optional[str] = None
    valor: float = 0.0
    competencia: Optional[str] = None
    parcelado: bool = False
    total_parcelas: Optional[int] = None
    parcela_atual: Optional[int] = None
    status: str = "pendente"          # pendente | pago
    observacoes: Optional[str] = None


class DespesaLocalUpdate(BaseModel):
    nome: Optional[str] = None
    categoria: Optional[str] = None
    subcategoria: Optional[str] = None
    valor: Optional[float] = None
    competencia: Optional[str] = None
    parcelado: Optional[bool] = None
    total_parcelas: Optional[int] = None
    parcela_atual: Optional[int] = None
    status: Optional[str] = None
    observacoes: Optional[str] = None


@router.get("")
def list_despesas_locais():
    despesas   = despesas_locais_storage.all()
    total      = sum(float(d.get("valor", 0)) for d in despesas)
    total_pago = sum(float(d.get("valor", 0)) for d in despesas if d.get("status") == "pago")
    total_pend = sum(float(d.get("valor", 0)) for d in despesas if d.get("status") != "pago")
    return {
        "despesas":       despesas,
        "total":          round(total, 2),
        "total_pago":     round(total_pago, 2),
        "total_pendente": round(total_pend, 2),
        "quantidade":     len(despesas),
    }


@router.post("")
def create_despesa_local(body: DespesaLocalCreate):
    return despesas_locais_storage.create(body.model_dump())


@router.put("/{id}")
def update_despesa_local(id: str, body: DespesaLocalUpdate):
    payload = {k: v for k, v in body.model_dump().items() if v is not None}
    updated = despesas_locais_storage.update(id, payload)
    if not updated:
        raise HTTPException(status_code=404, detail="Despesa não encontrada")
    return updated


@router.delete("/{id}")
def delete_despesa_local(id: str):
    if not despesas_locais_storage.delete(id):
        raise HTTPException(status_code=404, detail="Despesa não encontrada")
    return {"ok": True}

File: functions/src/routes/despesasLocais.ts
import { Router } from "express";
import { despesasLocaisStorage } from "../lib/firestore";

const router = Router();

function toFloat(v: unknown): number {
  const n = parseFloat(String(v ?? "0"));
  return isFinite(n) ? n : 0;
}

// GET /despesas-locais
router.get("/", async (_req, res) => {
  try {
    const items = await despesasLocaisStorage.all();
    const total = items.reduce((s, d) => s + toFloat(d.valor), 0);
    const total_pago = items.filter((d) => d.status === "pago").reduce((s, d) => s + toFloat(d.valor), 0);
    const total_pendente = items.filter((d) => d.status !== "pago").reduce((s, d) => s + toFloat(d.valor), 0);
    const r = (n: number) => Math.round(n * 100) / 100;
    res.json({ despesas: items, total: r(total), total_pago: r(total_pago), total_pendente: r(total_pendente), quantidade: items.length });
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

// POST /despesas-locais
router.post("/", async (req, res) => {
  try {
    res.status(201).json(await despesasLocaisStorage.create(req.body));
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

// PUT /despesas-locais/:id
router.put("/:id", async (req, res) => {
  try {
    const updated = await despesasLocaisStorage.update(req.params.id, req.body);
    if (!updated) { res.status(404).json({ detail: "Despesa não encontrada" }); return; }
    res.json(updated);
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

// DELETE /despesas-locais/:id
router.delete("/:id", async (req, res) => {
  try {
    if (!await despesasLocaisStorage.delete(req.params.id)) { res.status(404).json({ detail: "Despesa não encontrada" }); return; }
    res.json({ ok: true });
  } catch (e: unknown) {
    res.status(500).json({ detail: String(e) });
  }
});

export default router;
